### PR TITLE
Implement deriving strategies in th-desugar

### LINKED ISSUE #45: Support deriving strategies
See also #44.

### PR DESCRIPTION
**Pull request by RyanGlScott**
Fixes #45.

### CODE REVIEW

File: Language/Haskell/TH/Desugar/Core.hs [hunk 1]
```diff
@@ -233,9 +236,18 @@ data DInfo = DTyConI DDec (Maybe [DInstanceDec])
...
-data DDerivClause = DDerivClause (Maybe DerivStrategy) DCxt
-#endif
+-- | Corresponds to TH's @DerivClause@ type.
+data DDerivClause = DDerivClause (Maybe DDerivStrategy) DCxt
+                  deriving (Show, Typeable, Data, Generic)
+
+-- | Corresponds to TH's @DerivStrategy@ type.
+
+-- This is redefined here, since the Newtype constructor of DerivStrategy
+-- conflicts with the Newtype constructor of NewOrData.
```
Comment (RyanGlScott) [inline Language/Haskell/TH/Desugar/Core.hs]: I'm starting to regret naming that thing `Newtype` now... :)

File: .travis.yml [hunk 2]
```diff
@@ -5,7 +5,7 @@ sudo: required
 
 # The following enables several GHC versions to be tested; often it's enough to test only against the last release in a major GHC version. Feel free to omit lines listings versions you don't need/want testing for.
 env:
- - CABALVER=1.18 GHCVER=7.6.3
+ - CABALVER=1.24 GHCVER=7.6.3
```
Comment (goldfirere) [inline .travis.yml]: Why this change? Seems odd. Compatibility over a range of GHCs/cabals is important to th-desugar.

File: Language/Haskell/TH/Desugar.hs [hunk 3]
```diff
@@ -25,6 +25,7 @@ module Language.Haskell.TH.Desugar (
   -- * Desugared data types
   DExp(..), DLetDec(..), DPat(..), DType(..), DKind, DCxt, DPred(..),
   DTyVarBndr(..), DMatch(..), DClause(..), DDec(..),
+  DDerivClause(..), DDerivStrategy(..),
```
Comment (goldfirere) [inline Language/Haskell/TH/Desugar.hs]: We don't need `DDerivStrategy`, as we can just use TH's `DerivStrategy`.

Never mind. I see why you did this. Argh. But I don't have a better solution.

Haskell needs namespaces!

File: Language/Haskell/TH/Desugar/Core.hs [hunk 4]
```diff
@@ -797,14 +810,17 @@ dsDec (ClosedTypeFamilyD n tvbs m_k eqns) = do
 dsDec (RoleAnnotD n roles) = return [DRoleAnnotD n roles]
 #endif
 #if __GLASGOW_HASKELL__ >= 709
-#if __GLASGOW_HASKELL__ >= 801
-dsDec (StandaloneDerivD (Just _) cxt ty) = fail "Deriving strategies not supported by th-desugar."
-dsDec (StandaloneDerivD Nothing cxt ty) =
-#else
+# if MIN_VERSION_template_haskell(2,12,0)
```
Comment (goldfirere) [inline Language/Haskell/TH/Desugar/Core.hs]: Please be consistent with local style and use `#if`, without the space. Thanks!

File: Language/Haskell/TH/Desugar/Core.hs [hunk 5]
```diff
@@ -15,6 +15,9 @@ module Language.Haskell.TH.Desugar.Core where
 
 import Prelude hiding (mapM, foldl, foldr, all, elem, exp, concatMap, and)
 
+#if MIN_VERSION_template_haskell(2,12,0)
```
Comment (goldfirere) [inline Language/Haskell/TH/Desugar/Core.hs]: Please use `#if __GLASGOW_HASKELL__ >= 801` (here and elsewhere). That way, I can compile without cabal.

File: Language/Haskell/TH/Desugar/Core.hs [hunk 6]
```diff
@@ -1043,13 +1059,23 @@ dsTvb (KindedTV n k) = DKindedTV n <$> dsType k
...
-dsDerivClause (DerivClause Nothing cxt) = dsCxt cxt
-dsDerivClause (DerivClause _ cxt) = fail "Deriving strategies not supported in th-desugar."
+#if MIN_VERSION_template_haskell(2,12,0)
+-- | Desugar a @DerivClause@.
+dsDerivClause :: DsMonad q => DerivClause -> q DDerivClause
+dsDerivClause (DerivClause mds cxt) =
+  DDerivClause (fmap dsDerivStrategy mds) <$> dsCxt cxt
+
+-- | Desugar a @DerivStrategy@.
+dsDerivStrategy :: DerivStrategy -> DDerivStrategy
```
Comment (goldfirere) [inline Language/Haskell/TH/Desugar/Core.hs]: Should this function be exported from `Language.Haskell.TH.Desugar`? I suppose so.

File: Test/Run.hs [hunk 7]
```diff
@@ -201,6 +201,13 @@ test_standalone_deriving = (MkBlarggie 5 'x') == (MkBlarggie 5 'x')
 test_standalone_deriving = True
 #endif
 
+test_deriving_strategies :: Bool
```
Comment (goldfirere) [inline Test/Run.hs]: Thanks for adding this test!

File: Language/Haskell/TH/Desugar.hs (hunk 3 above)
Comment (RyanGlScott) [inline Language/Haskell/TH/Desugar.hs]: I have considered going back into GHC and changing `Newtype` to `NewtypeStrategy`, `Stock` to `StockStrategy`, and `Anyclass` to `AnyclassStrategy`. That would be one way around this :)

File: .travis.yml (hunk 2 above)
Comment (goldfirere) [inline .travis.yml]: OK. Not worth the energy to understand better. Thanks.

File: Language/Haskell/TH/Desugar.hs (hunk 3 above)
Comment (goldfirere) [inline Language/Haskell/TH/Desugar.hs]: I'd actually be in favor of this resolution. Deriving strategies are obscure (but necessary -- not meant to diminish your work in implementing them!), and `Newtype` seems to take up a prime spot in the namespace.

File: Language/Haskell/TH/Desugar/Core.hs (hunk 6 above)
Comment (RyanGlScott) [inline Language/Haskell/TH/Desugar/Core.hs]: Good idea. I also added a note to the Haddocks explaining that it's 8.2+ only.

File: Language/Haskell/TH/Desugar.hs (hunk 3 above)
Comment (RyanGlScott) [inline Language/Haskell/TH/Desugar.hs]: Cool, I can prepare a patch that does that.

I suppose we could put this PR on hiatus until that gets merged, or we could adopt this change now and change to the new names later. Up to you.

**Comment by RyanGlScott** [comment]
@goldfirere: Now that http://git.haskell.org/ghc.git/commit/5349d648fd7af3f50953e8594b3d148ab073017f is in GHC HEAD, I've updated this PR so that we reuse the `DerivStrategy` datatype from TH. Much cleaner!

One small problem is that the testsuite doesn't compile on GHC HEAD:

```
[4 of 4] Compiling Run              ( Test/Run.hs, dist/build/spec/spec-tmp/Run.o )

Test/Run.hs:134:14: error:
    • Couldn't match expected type ‘PolyTF IO’ with actual type ‘Bool’
    • In the pattern: True
      In an equation for ‘f_aMl3’: f_aMl3 True = ()
      In the expression:
        (let
           f_aMl3 :: PolyTF IO -> ()
           f_aMl3 True = ()
         in f_aMl3)
```

And whaddya know, turns out [this is my fault](https://github.com/goldfirere/th-desugar/blob/2ced2834e3e313970b9d029f87995cf0fa86adcf/Test/Run.hs#L134-L136) too :)

```haskell
test_e8a = $(test_expand8 >>= dsExp >>= expand >>= return . expToTH)
  -- the line above should fail once GHC#8953 is fixed for closed type
  -- families
```

What should we do with this test now that GHC Trac #8953 (and thus #43) has been fixed? Conditionally include it with CPP? Remove it? Something else?

**Comment by goldfirere** [comment]
What is that test doing? Is there a way to fix it? (A faint memory says: no.) If not, then just disable it in HEAD. And, now that you've brought us up-to-date with HEAD, enable testing of HEAD with Travis, please.

Many thanks for all this work -- much appreciated!

**Comment by RyanGlScott** [comment]
> What is that test doing? Is there a way to fix it? (A faint memory says: no.) If not, then just disable it in HEAD.

It seems to be testing the (faulty) type expansion of [this particular closed type family](https://github.com/goldfirere/th-desugar/blob/7e0aca246ecd63bd7c1ca75bea0332d4d15690aa/Test/Splices.hs#L247-L248):

```haskell
type family PolyTF (x :: k) :: * where
  PolyTF (x :: *) = Bool
```

Now that GHC Trac #8953 has been fixed, I'd argue that it no longer makes sense. I've disabled it `#if __GLASGOW_HASKELL__ >= 801`.

> And, now that you've brought us up-to-date with HEAD, enable testing of HEAD with Travis, please.

Done. Sadly, I encountered some more failing tests in the testsuite:

```
Failures:

  Test/Run.hs:264: 
  1) th-desugar library passes dec test 9

  Test/Run.hs:295: 
  2) th-desugar library reifies local definition 19
```

I'll have to investigate further to see what's wrong.

**Comment by RyanGlScott** [comment]
OK, I see why `th-desugar library reifies local definition 19` is failing. Consider this type family from the `th-desugar` test suite:

```haskell
type family R21 (a :: k) (b :: k) :: k where R21 a b = b
```

If you reify it normally, you get kind signatures on its LHS types, as expected:

```
$ inplace/bin/ghc-stage2 --interactive
GHCi, version 8.1.20161208: http://www.haskell.org/ghc/  :? for help
Loaded GHCi configuration from /home/rgscott/.ghci
λ> :set -XKindSignatures -XPolyKinds -XTypeFamilies -XTemplateHaskell
λ> import Language.Haskell.TH
λ> type family R21 (a :: k) (b :: k) :: k where R21 a b = b
λ> putStrLn $(reify ''R21 >>= stringE . pprint)
type family Ghci1.R21 (a_0 :: k_1) (b_2 :: k_1) :: k_1 where
    Ghci1.R21 (a_3 :: k_4) (b_5 :: k_4) = b_5
```

But if you quote it, you won't get the kind signatures!

```
λ> putStrLn $([d| type family R21 (a :: k) (b :: k) :: k where R21 a b = b |] >>= stringE . pprint)type family R21_0 (a_1 :: k_2) (b_3 :: k_2) :: k_2 where
    R21_0 a_4 b_5 = b_5
```

Would it be cheating to fix this test by just giving explicit kind signatures to `a` and `b` in `R21 a b = b`?

**Comment by RyanGlScott** [comment]
Following your advice in https://github.com/goldfirere/th-desugar/pull/48#issuecomment-266778976, I was able to fix `th-desugar library reifies local definition 19`! And weirdly enough, I'm no longer able to trigger `th-desugar library passes dec test 9`, so I believe I've fixed all the necessary tests.

Other than the Travis build for GHC HEAD not working satisfactorily (as noted in https://github.com/goldfirere/th-desugar/pull/46#issuecomment-266575931), this PR should be ready to be merged, pending any further suggestions.

**Comment by goldfirere** [comment]
Hooray!
